src/main/kotlin/com/safehill/kclient/tasks/BackgroundTaskProcessor.kt

```diff
@@ -1,66 +1,124 @@
 package com.safehill.kclient.tasks
 
 import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.delay
 import kotlinx.coroutines.isActive
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
 import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicLong
+import kotlin.coroutines.coroutineContext
 import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
 
-class BackgroundTaskProcessor<T : BackgroundTask>(
-    processorScope: CoroutineScope,
-    private val jobScope: CoroutineScope
-) {
-
-    internal val taskQueue = ConcurrentLinkedQueue<T>()
-    private val processingMutex = Mutex()
-    private var currentJob: Job? = null
-    private var repeatingLifecycle: Job? = null
-
-    init {
-        processorScope.launch {
-            while (isActive) {
-                if (currentJob?.isActive != true && taskQueue.isNotEmpty()) {
-                    processTasks()
+
+class BackgroundTaskProcessor<T : BackgroundTask> {
+
+    private val taskCounter = AtomicLong(0)
+
+    private val mutex = Mutex()
+
+    private val jobQueue = ConcurrentLinkedQueue<JobWithCounter>()
+
+    suspend fun run(
+        task: T,
+        enqueueMode: EnqueueMode = EnqueueMode.DropLatest,
+        repeatMode: RepeatMode = RepeatMode.Once
+    ) {
+        do {
+            try {
+                coroutineScope {
+                    val jobId: Long = mutex.withLock {
+                        enqueueMode.takeCorrespondingActions()
+                        addTaskToQueue(task)
+                    }
+                    waitForCompletion(jobId)
                 }
+            } catch (_: TaskAlreadyOnQueue) {
             }
+            delay(repeatMode.getWaitDuration())
+        } while (coroutineContext.isActive && repeatMode is RepeatMode.Repeating)
+    }
+
+    private fun CoroutineScope.addTaskToQueue(task: T): Long {
+        val jobId = taskCounter.incrementAndGet()
+        val job = launch(
+            start = CoroutineStart.LAZY
+        ) {
+            task.run()
+        }
+        jobQueue.add(job with jobId)
+        job.invokeOnCompletion {
+            jobQueue.removeIf { it.counter == jobId }
+        }
+        return jobId
+    }
+
+    private fun dropJobsInQueue() {
+        while (jobQueue.isNotEmpty()) {
+            jobQueue.poll()?.job?.cancel()
```

Comment:
Yeah.. the task that was added gets cancelled if the coroutine that started it gets cancelled.

Reply:
Right, but my question is if there's an operation in progress it's up to the task to check whether it's been cancelled right? Otherwise it'll finish first

Reply:
We can control that we EnqueueMode.
if the EnqueueMode is Enqueue, it is just gonna wait for the current task to finish and then start the new task.
if its DropOngoing, it is gonna drop all the pending tasks in the queue and start the newly added task.
if its DropLatest, its gonna drop the new task if one is already running. If not, the new task starts.
DropLatest is the default enqueue model.

Reply:
Added comment for more explanation.

Reply:
> Are the current task implementations responsive to canceling?

Oh... I get what you mean... Coroutines check for cancellation by default and the current implementations are cancellation cooperative.

Reply:
I should not say "by default" tbh. Most of the library functions support cancellation. The socket connection support cancellation as well. Fuel supports coroutine cancellation as well. As far as I know, our implementations are cooperative as well.